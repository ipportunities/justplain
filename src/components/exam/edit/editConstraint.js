import React from 'react'
import { useSelector } from 'react-redux'
import apiCall from '../../api'

export const EditConstraint = ({ categoryId, subject, difficulty, amount, setAmount, updateMaxPointRange }) => {
  const [maxAmountOfQuestions, setMaxAmountOfQuestions] = React.useState(0)
  const auth = useSelector(state => state.auth)
  const [maxPointsRange, setMaxPointsRange] = React.useState([0, 0])

  React.useEffect(() => {
    apiCall({
      action: 'get_max_amount_of_questions_for_constraint',
      token: auth.token,
      data: { categoryId: categoryId, subject: subject, difficulty: difficulty }
    }).then(resp => {
      if (+resp.maxAmountOfQuestions < amount) {
        setAmount(+resp.maxAmountOfQuestions)
      }
      setMaxAmountOfQuestions(resp.maxAmountOfQuestions)
    })
  }, [subject, difficulty])

  React.useEffect(() => {
    if (amount > 0) {
      apiCall({
        action: 'get_max_points_range',
        token: auth.token,
        data: { categoryId: categoryId, subject: subject, difficulty: difficulty, amountOfQuestions: amount }
      }).then(resp => {
        updateMaxPointRange(`${categoryId}-${subject}-${difficulty}`, resp.maxPointsRange)
        setMaxPointsRange(resp.maxPointsRange)
      })
    }
  }, [subject, difficulty, amount])

  const setNewAmount = (value) => {
    setAmount(value <= maxAmountOfQuestions ? value : maxAmountOfQuestions)
  }

  return <div className="row m-0">
    <input type="number" className="form-control col-4 pl-2 pr-2" id="AmountOfQuestionsRange"
           onChange={ e => setNewAmount(+e.target.value) } min={ 0 } max={ maxAmountOfQuestions }
           name="amountOfQuestions" value={ amount } disabled={ +maxAmountOfQuestions === 0 } />
    <div className="col">{ `/ ${ maxAmountOfQuestions } (${ maxPointsRange[0] }-${ maxPointsRange[1] } pnts)` }</div>
  </div>
}