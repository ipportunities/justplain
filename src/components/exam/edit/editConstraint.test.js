import React from 'react'
import { shallow } from 'enzyme'
import { EditConstraint } from './editConstraint'

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn().mockReturnValue({ auth: { token: 'a' } })
}))

const saveMock = jest.fn()
const initialConstraintMock = { id: 8, amountOfQuestions: 10, difficulty: 'EASY', subject: '' }

describe('Exam Type Constraint Edit', () => {
  it('should render correctly', () => {
    const component = shallow(<EditConstraint initialConstraint={ initialConstraintMock }
                                              subjects={ ['Theoretical', 'Practical'] }
                                              difficulties={ ['EASY', 'HARD'] } className='col-10 pt-4'
                                              maxAmountOfQuestions={ 15 } save={ saveMock } />)
    expect(component).toMatchSnapshot()
  })
})
