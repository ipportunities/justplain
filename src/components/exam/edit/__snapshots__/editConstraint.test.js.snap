// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exam Type Constraint Edit should render correctly 1`] = `
<div
  className="row m-0"
>
  <input
    className="form-control col-4 pl-2 pr-2"
    disabled={true}
    id="AmountOfQuestionsRange"
    max={0}
    min={0}
    name="amountOfQuestions"
    onChange={[Function]}
    type="number"
  />
  <div
    className="col"
  >
    / 0 (0-0 pnts)
  </div>
</div>
`;
